Change equivalence class hashing function

misterikkit opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This updates the Pod equivalence class hashing function to hash pod fields which are read by scheduler predicates. Until now, we used a pod's OwnerReference as a shorthand for equivalence, but not all controllers will create homogeneous sets of pods.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
/ref #58222 

**Performance impact**:
Hashing is not expensive enough to impact scheduling performance.

|Test|Result|
|---|---|
| Before | `BenchmarkEquivalenceHash-40       200000              7722 ns/op` |
| After | `BenchmarkEquivalenceHash-40        10000            114184 ns/op` |

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```
/sig scheduling

Commit:
Change equivalence hash function.

This changes the equivalence class hashing function to use as inputs all
the Pod fields which are read by FitPredicates. Before we used a
combination of OwnerReference and PersistentVolumeClaim info, which was
a close approximation. The new method ensures that hashing remains
correct regardless of controller behavior.

The PVCSet field can be removed from equivalencePod because it is
implicitly included in the Volume list.

Tests are now broken.

Commit:
Move equivalence class hash code.

This moves the equivalence hashing code from
algorithm/predicates/utils.go to core/equivalence_cache.go.

In the process, making the hashing function and hashing function factory
both injectable dependencies is removed.

misterikkit commented:
I believe the change is correct right now, but I still need to add new tests. I also plan to simplify the hashing code in this PR.

resouer commented:
/assign resouer

resouer commented:
/ok-to-test

misterikkit commented:
Regarding pod equivalence hash, I have this unanswered question.

Should we hash PersistentVolumeClaimVolumeSource objects if they name a PersistentVolumeClaim that doesn't exist? Would it be incorrect to hash them anyway?

The current implementation skips over any PVC volume sources if their name doesn't match a PVC object. We don't check whether the claim is bound, just whether it exists.

I would prefer to avoid calling out to the client-go API to check whether the PVC exists.

bsalamat commented:
> Should we hash PersistentVolumeClaimVolumeSource objects if they name a PersistentVolumeClaim that doesn't exist? Would it be incorrect to hash them anyway?

I think it would be fine to cache them regardless. @resouer, correct me if I am wrong.

misterikkit commented:
I discussed this with a few people, and the conclusion is that we can safely ignore the PVC informer when computing the equivalence hash of a pod.

Based on that, I will add a patch to this PR which removes the informer reference from the equivalence hashing code.

Commit:
Add benchmark for equivalence hashing.

resouer commented:
> Should we hash PersistentVolumeClaimVolumeSource objects if they name a PersistentVolumeClaim that doesn't exist? Would it be incorrect to hash them anyway?

Yes, it would be fine to hash it regardless. While I need to take a look at the latest commit :)

resouer commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -208,15 +206,47 @@ func (ec *EquivalenceCache) InvalidateCachedPredicateItemForPodAdd(pod *v1.Pod,
+// equivalencePod should be normalized. (e.g. by sorting them) However, the
+// vast majority of equivalent pod classes are expected to be created from a
+// single pod template, so they will all have the same ordering.
+type equivalencePod struct {
```
Any reason why PVC is removed?

resouer commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -208,15 +206,47 @@ func (ec *EquivalenceCache) InvalidateCachedPredicateItemForPodAdd(pod *v1.Pod,
+// getEquivalencePod returns the equivalencePod for a Pod.
+func getEquivalencePod(pod *v1.Pod) *equivalencePod {
+	return &equivalencePod{
+		Namespace:      &pod.Namespace,
```
It seems the namspaced resource is used only in: Vlume related predicates, Service Affinity, Pod affinity/anti-affinity (correct me if I made mistake). 

So do you think it make sense that we calculate hash for `namespace + Affinity` and `namespace + Volume` separately, then in most cases equiv cache can be re-used across namespaces?

resouer replied:
Service Affinity only influenced by Pod label (for selector). It's fine to leave it in hash calculation.

bsalamat replied:
> It seems the namspaced resource is used only in: Volume related predicates, Service Affinity, Pod affinity/anti-affinity (correct me if I made mistake).
So do you think it make sense that we calculate hash for namespace + Affinity and namespace + Volume separately, then in most cases equiv cache can be re-used across namespaces?

I am not sure if I understand. Are you saying that we don't need to have `namespace` here?

misterikkit replied:
I removed PVCSet because all PVCVolumeInfo objects are in the Volumes list. Based on my understanding of that struct, it is implied to have the same namespace as the pod. i.e. the same claim name could appear in two different namespaces.


If I understand Harry's other comment, he's saying that we should conditionally hash namespace based on whether Affinity or Volume is present. That logic seems valid to me, although the benefit is probably marginal. I will try making the change, but if it makes the code a lot messier, then I'd prefer we keep it as is.

resouer replied:
Yes, not sure if it's practical but would like to discuss.

bsalamat replied:
> it make more sense if equiv class cache can be applied/reused by pods across all namespaces.

I think your point is valid, but I guess the improvement would not be significant as Jonathan said.

resouer replied:
Agreed.

resouer commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -27,6 +27,51 @@ import (
 )
 
+// makeBasicPod returns a Pod object with many of the fields populated.
+func makeBasicPod() *v1.Pod {
```
More unit tests would be valuable.

misterikkit replied:
Working on it (=

misterikkit commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -208,15 +206,47 @@ func (ec *EquivalenceCache) InvalidateCachedPredicateItemForPodAdd(pod *v1.Pod,
+// The hashing function returns the same value for any two pods that are
+// equivalent from the perspective of scheduling.
+func (ec *EquivalenceCache) getEquivalenceHash(pod *v1.Pod) (uint64, bool) {
+	equivalencePod := getEquivalencePod(pod)
```
Sure, I can add a benchmark for this.

Commit:
Fix equivalence cache hash tests.

misterikkit commented:
I've rebased and cleaned up. Although I think it could still use more tests.

misterikkit commented:
Integration tests are currently broken because pods that had no OwnerReference previously are now hashing to the same value. We suspect the cache is not being invalidated for the specific case under test.

resouer commented:
I saw a crash in: 
```
k8s.io/kubernetes/test/integration/scheduler TestVolumeBinding 0.00s
go test -v k8s.io/kubernetes/test/integration/scheduler -run TestVolumeBinding$
```
I would guess change the equivalence cache factory caused nil pointer in that test.

Commit:
Fix equiv. cache invalidation of Node condition.

Equivalence cache for CheckNodeConditionPred becomes invalid when
Node.Spec.Unschedulable changes. This can happen even if
Node.Status.Conditions does not change, so move the logic around.

This logic is covered by integration test
"test/integration/scheduler".TestUnschedulableNodes but equivalence
cache is currently skipped when test pods have no OwnerReference.

misterikkit commented:
> I would guess change the equivalence cache factory caused nil pointer in that test.

When I read the output for that failed test, it just looks like a list of every source file in k8s.io/kubernetes. Is this a signal that the test hit a segfault?

The test passes for me on my local machine, so I'm not sure how to debug the failure. Maybe it's a flake?

misterikkit commented:
I've rebased on #58780 which should fix some of the broken tests.
/retest

bsalamat commented:
/lgtm

Thanks, @misterikkit!

liggitt commented:
a large number of scheduling flakes started appearing at the time this PR was merged:

https://storage.googleapis.com/k8s-gubernator/triage/index.html?ci=0&pr=1&job=pull-kubernetes-unit&test=k8s.io%2Fkubernetes%2Ftest%2Fintegration%2Fscheduler

@bsalamat @misterikkit can you triage whether this PR was the root cause and determine if there's a fix or if we need to revert?

bsalamat commented:
@resouer Now that the hash is calculated once per schedule, I think we should resubmit this PR. WDYT?

resouer commented:
@bsalamat Yep, and this time we need to be more cautious to see if any integration tests failure happen. 

So we need to send a revert revert PR? Or draft a new one?

bsalamat commented:
> So we need to send a revert revert PR? Or @misterikkit will draft a new one?

Yes, we should send a revert of the revert, which is going to be the same as this PR.
Jonathan is on paternity leave and won't be back for a couple of months or so.

resouer commented:
OK, I'll send a PR to see if integration tests are satisfied.

resouer commented:
@bsalamat Oh, github does not allow revert revert due to something has been changed (should be the the "hash is calculated once per schedule"  PR)

I may have to create a new one based on this, I think @misterikkit will be fine with it.

bsalamat commented:
@resouer That would be fine. Please refer to this PR and mention that your are resending it. Thanks!

resouer commented:
I probably found out the root cause of timeout, see comments here: https://github.com/kubernetes/kubernetes/pull/60753#issuecomment-370915052

I will revisit the logic and try to fix it.
